src/main/java/org/folio/circulation/resources/PickSlipsReportResource.java

```diff
@@ -0,0 +1,199 @@
+      .thenComposeAsync(r -> r.after(items -> fetchHoldings(items, clients.holdingsStorage())))
+      .thenComposeAsync(r -> r.after(items -> fetchInstances(items, clients.instancesStorage())))
+      .thenApply(r -> r.next(this::mapResultToJson))
+      .thenApply(OkJsonResponseResult::from)
+      .thenAccept(r -> r.writeTo(routingContext.response()));
+  }
+
+  private CompletableFuture<Result<Set<Location>>> fetchLocationsForServicePoint(
+    UUID servicePointId, CollectionResourceClient locationsStorageClient) {
+
+    return new MultipleRecordFetcher<>(locationsStorageClient, LOCATIONS_KEY, Location::from)
+      .findByQuery(exactMatch(PRIMARY_SERVICE_POINT_KEY, servicePointId.toString()))
+      .thenApply(r -> r.next(this::recordsToSet));
+  }
+
+  private CompletableFuture<Result<Set<Item>>> fetchPagedItemsByLocations(
+    Set<Location> locations, CollectionResourceClient itemStorageClient) {
+
+    Set<String> locationIds = locations.stream()
+      .map(Location::getId)
+      .collect(Collectors.toSet());
+
+    final Result<CqlQuery> statusQuery = exactMatch(STATUS_NAME_KEY, ItemStatus.PAGED.getValue());
+
+    return new MultipleRecordFetcher<>(itemStorageClient, ITEMS_KEY, Item::from)
+      .findByIndexNameAndQuery(locationIds, EFFECTIVE_LOCATION_ID_KEY, statusQuery)
+      .thenApply(r -> r.next(this::recordsToSet))
+      .thenApply(r -> r.next(items -> populateLocationsInItems(items, locations)));
+  }
+
+  private Result<Set<Item>> populateLocationsInItems(Set<Item> items, Set<Location> locations) {
+    return Result.succeeded(
+      items.stream()
+        .map(item -> item.withLocation(
+          locations.stream()
+            .filter(location -> location.getId().equals(item.getLocationId()))
+            .findFirst()
+            .orElse(null)))
+        .collect(Collectors.toSet()));
+  }
+
+  private CompletableFuture<Result<Set<Item>>> filterItemsByOpenUnfilledRequests(
+      Set<Item> items, CollectionResourceClient client) {
+
+    Set<String> itemIds = items.stream()
+        .map(Item::getItemId)
+        .collect(Collectors.toSet());
+
+    final Result<CqlQuery> statusQuery = exactMatch(STATUS_KEY, RequestStatus.OPEN_NOT_YET_FILLED.getValue());
+
+    return new MultipleRecordFetcher<>(client, REQUESTS_KEY, Request::from)
+      .findByIndexNameAndQuery(itemIds, ITEM_ID_KEY, statusQuery)
+      .thenApply(r -> r.next(this::recordsToSet))
+      .thenApply(r -> r.next(requests -> filterItemsByRequests(items, requests)));
+  }
+
+  private Result<Set<Item>> filterItemsByRequests(Set<Item> items, Set<Request> requests) {
+    Set<String> requestedItemIds = requests.stream()
+        .map(Request::getItemId)
+        .collect(Collectors.toSet());
+
+    return Result.succeeded(items.stream()
+      .filter(i -> requestedItemIds.contains(i.getItemId()))
+      .collect(Collectors.toSet()));
+  }
+
+  private CompletableFuture<Result<Set<Item>>> fetchHoldings(
```

Comment:
I don't see the place where we need holdings and instances, could you please double-check if they required?

Reply:
I need `title` and `contributors` of an item in the response. Correct me if I'm wrong, but these properties come from `instance`, and to fetch `instance` we need `instanceId` which in turn comes from `holdings`.

Reply:
Thanks @OleksandrVidinieiev, you're right.